File: 2015/07.js
import { graph, expr } from './utils'

const processData = (acc, l) => {
  const ins = l.scan(/[a-z]+|\d+|[A-Z]+/g)
  const names = l.scan(/[a-z]+/g)
  const name = names.pop()
  acc.map[name] = names
  acc.eqn[name] = ins.remove(name)
  return acc
}

const prepare = data => {
  const tmp = data.reduce(processData, { 
    map: {}, eqn: {}, val: {},
    graph: null
  })
  tmp.graph = graph(tmp.map)
  return tmp
}
const solve = data => data.graph
  .tsort()
  .reduce((acc, d) => {
    acc.val[d] = expr(acc.eqn[d], acc.val)
    return acc
  }, data)
const part1 = solve
const part2 = data => {
  data.eqn.b[0] = `${data.val.a}`
  return solve(data)
}
const finish = obj => obj.val.a

export default () => ({ prepare, part1, part2, finish })

File: 2015/rb/07.rb
fn = open("2015/07.input")

Deps = Hash.new
input = fn.each_line.map(&:chomp).map{|l|
  ins = l.scan(/[a-z]+|\d+|[A-Z]+/)
  names = l.scan(/[a-z]+/)
  name = names.pop
  Deps[name] = [names, ins - [name]]
}

require_relative './utils/expr'
def solve()
  found = []
  vals = {}
  while found.size < Deps.size
    selected = Deps.select{ |k, v| v[0] & found == v[0]}.keys - found  
    selected.each{|name| 
      found << name
      names, eqn = Deps[name]
      vals[name] = expr(eqn, vals)
    }
  end
  vals
end

# Part 1
sln = solve()
p sln['a']

# Part 2
Deps['b'][1][0] = sln['a'].to_s
sln = solve()
p sln['a']